modules/cnf-troubleshooting-numa-aware-workloads.adoc

```diff
@@ -0,0 +1,113 @@
+// Module included in the following assemblies:
+//
+// *scalability_and_performance/cnf-topology-aware-scheduling.adoc
+
+[id="cnf-troubleshooting-numa-aware-workloads_{context}"]
+= Troubleshooting topology-aware workload scheduling
+
+Troubleshoot common problems with topology-aware workload scheduling by performing the following steps.
+
+.Prerequisites
+
+. Install the {product-title} CLI (oc).
+. Log in afor ts a user with cluster-admin privileges.
+
+.Procedure
+
+. Verify that the `noderesourcetopologies` CRD is deployed in the cluster:
++
+[source,terminal]
+----
+$ oc get crd | grep noderesourcetopologies
+----
++
+.Example output
+[source,terminal]
+----
+NAME                                                              CREATED AT
+noderesourcetopologies.topology.node.k8s.io                       2022-01-18T08:28:06Z
+----
+
+. Check that the topology-aware scheduler name matches the name specified in your topology-aware workloads:
++
+[source,terminal]
+----
+$ oc get numaresourcesschedulers.nodetopology.openshift.io numaresourcesscheduler -o json | jq '.status.schedulerName'
+----
++
+.Example output
+[source,terminal]
+----
+topo-aware-scheduler
+----
+
+. Verify that topology-aware scheduable nodes have the `noderesourcetopologies` CR applied to them.
++
+[source,terminal]
+----
+$ oc get noderesourcetopologies.topology.node.k8s.io
+----
++
+.Example output
+[source,terminal]
+----
+NAME                    AGE
+compute-0.example.com   17h
+----
+
+. Verify the NUMA zone granularity for all scheduable nodes:
++
+[source,terminal]
+----
+$ oc get noderesourcetopologies.topology.node.k8s.io -o yaml
+----
++
+.Example output
+[source,terminal]
+----
+apiVersion: v1
+items:
+- apiVersion: topology.node.k8s.io/v1alpha1
+  kind: NodeResourceTopology
+  metadata:
+    annotations:
+      k8stopoawareschedwg/rte-update: periodic
+    creationTimestamp: "2022-01-24T18:13:01Z"
+    generation: 2
+    name: compute-0.example.com
+    resourceVersion: "5415846"
+    uid: eae6d6ce-77a0-4198-8671-89aafa691e58
+  topologyPolicies:
+  - SingleNUMANodeContainerLevel
+  zones:                   <1>
+  - costs:
+    - name: node-0
+      value: 10
+    - name: node-1
+      value: 20
+    name: node-0
+    resources:             <2>
+    - allocatable: "15"
+      available: "5"
+      capacity: "16"
+      name: cpu
+    type: Node
+  - costs:
+    - name: node-0
+      value: 20
+    - name: node-1
+      value: 10
+    name: node-1
+    resources:
+    - allocatable: "15"
+      available: "7"
+      capacity: "16"
+      name: cpu
+    type: Node
+kind: List
+metadata:
+  resourceVersion: ""
+  selfLink: ""
+----
+<1> Each stanza under `zones` describes the resources for a single NUMA zone.
+<2> `resources` describes the current state of the NUMA zone resources.
```

Comment:
let's add a note that as troubleshooting, we may want to check that available resources are properly reflected in the values reported here (`available`).
The only way to check is unfortunately to do the math manually and compare the value with what is reported; resources tracked here are only these allocated to guaranteed pods (NOT what is allocated to burstable pods).
Unfortunately I can't suggest a good wording to express this

Reply:
in order to troubleshoot updates (or lack thereof) to noderesourcetopologies objects, we need to inspect the relevant resourcetopologyexporter container logs. Thus we need to provide instructions about how to find the right pod to inspect knowing which noderesourcetopology instance we need to debug. I'll provide an example flow later, or perhaps we can defer to a followup update/KB article?

Reply:
last, in order to troubleshoot scheduling decisions the best shot is to check the numa aware scheduler logs. In this case there is just a single istance, so it's easier. I'll provide a example flow later, but in general
- if needed, increase the scheduler plugin log level. We can do so using a field in the `NUMAResourcesScheduler` object, but this will require removing the plugin/redploying it. This may cause a small but non-null service availability disruption - existing workload WILL keep running ofcourse, but no new workload can be scheduled in between
- check the numa aware scheduler logs using the standard `oc logs ...` flow - example pending

Reply:
to configure the log level of the scheduler:
set the `logLevel` field. Valid values are as follows:
"Normal" is the default.  Normal, working log information, everything is fine, but helpful notices for auditing or common operations.
"Debug" is used when something went wrong.  Even common operations may be logged, and less helpful but more quantity of notices.
"Trace" is used when something went really badly and even more verbose logs are needed.  Logging every function call as part of a common operation, to tracing execution of a query.

example:
```
apiVersion: nodetopology.openshift.io/v1alpha1
kind: NUMAResourcesScheduler
metadata:
  name: numaresourcesscheduler
spec:
  imageSpec: "repo.io/awesome-organization/awesome-scheduler:devel"
  logLevel: "Trace"
```

worth stressing we cannot (yet) edit in place: better to delete the existing CR and to submit a new one. The downtime is expected to be minimal, with the biggest offender being the scheduler image download time.

Reply:
to get the scheduler plugin logs:
1. get the deployment information, reported by the operator in the `status` field of the `numaresourcesscheduler` object. This is just the simplest way to learn about the pod name, you can also inspect manually and learn the same thing in a different way: ```
 $ oc get numaresourcesschedulers.nodetopology.openshift.io numaresourcesscheduler -o json | jq '.status.deployment'
{
  "name": "secondary-scheduler",
  "namespace": "numaresources"
}```
2. (standard kube procedure) get the label the deployment found in the step above use to refer to the pods created by it ```$ oc get deployment -n numaresources secondary-scheduler -o json | jq .metadata.labels```
3. get pod name using the labels: ```$ oc get pods -n numaresources -l'app=secondary-scheduler'
NAME                                   READY   STATUS    RESTARTS   AGE
secondary-scheduler-6cc8fdcdf6-hrp78   1/1     Running   0          95m```
4. now we know namespace and pod name, we can finally run: ```oc logs -n numaresources secondary-scheduler-6cc8fdcdf6-hrp7```

Reply:
> let's add a note that as troubleshooting, we may want to check that available resources are properly reflected in the values reported here (`available`). The only way to check is unfortunately to do the math manually and compare the value with what is reported; resources tracked here are only these allocated to guaranteed pods (NOT what is allocated to burstable pods). Unfortunately I can't suggest a good wording to express this

OK. How do we compare? is this something like: `oc debug node/<node_name>` and then `numactl --show` command?

Reply:
> OK. How do we compare? is this something like: oc debug node/<node_name> and then numactl --show command?

The TL;DR: version is we need to check the exclusive resources allocated to each guaranteed pod running on the container. So yes, `numactl` and/or `lscpu` will be needed, among other steps.
 This is complex to explain, we may want to defer to a later KB.